File: app/src/main/java/cc/wenmin92/androidlearn/MyApplication.java
package cc.wenmin92.androidlearn;

import android.app.Application;
import android.support.annotation.NonNull;
import android.util.Log;

import com.squareup.leakcanary.LeakCanary;

import timber.log.Timber;

public class MyApplication extends Application {

    @Override
    public void onCreate() {
        super.onCreate();
        initLogTool();
        initLeakCanary();
    }

    /**
     * 内存泄漏检查
     */
    private void initLeakCanary() {
        if (LeakCanary.isInAnalyzerProcess(this)) {
            return;
        }
        LeakCanary.install(this);
    }

    /**
     * 设置log工具
     */
    private void initLogTool() {
        if (BuildConfig.DEBUG)
            Timber.plant(new ThreadAwareDebugTree());
        else
            Timber.plant(new ReleaseTree());
    }

    public class ThreadAwareDebugTree extends Timber.DebugTree {
        @Override
        protected void log(int priority, String tag, @NonNull String message, Throwable t) {
            if (tag != null) {
                String threadName = Thread.currentThread().getName();
                tag = "<" + threadName + "> " + tag;
            }
            super.log(priority, tag, message, t);
        }

        @Override
        protected String createStackElementTag(@NonNull StackTraceElement element) {
            return super.createStackElementTag(element) + "(Line " + element.getLineNumber() + ")";  //日志显示行号
        }
    }

    public class ReleaseTree extends ThreadAwareDebugTree {
        @Override
        protected boolean isLoggable(String tag, int priority) {
            return priority != Log.VERBOSE && priority != Log.DEBUG && priority != Log.INFO;
        }

        @Override
        protected void log(int priority, String tag, @NonNull String message, Throwable t) {
            if (!isLoggable(tag, priority)) return;
            super.log(priority, tag, message, t);
        }
    }
}

File: app/src/main/java/cc/wenmin92/androidlearn/MyApplication.kt
package cc.wenmin92.androidlearn

import android.app.Application
import android.util.Log

import cn.leancloud.AVOSCloud
import timber.log.Timber

class MyApplication : Application() {

    override fun onCreate() {
        super.onCreate()
        initLogTool()
        initLeakCanary()
        initLeanCloud()
    }

    /**
     * 内存泄漏检查
     */
    private fun initLeakCanary() {
        // if (LeakCanary.isInAnalyzerProcess(this)) {
        //     return;
        // }
        // LeakCanary.install(this);
    }

    /**
     * 设置log工具
     */
    private fun initLogTool() {
        if (BuildConfig.DEBUG)
            Timber.plant(ThreadAwareDebugTree())
        else
            Timber.plant(ReleaseTree())
    }

    private fun initLeanCloud() {
        AVOSCloud.initialize(this, "dc3UTbiJwju2a177eoesMcXA-gzGzoHsz", "WOqp3lgj1xdW4TSi9AKhfzFr")
    }

    open inner class ThreadAwareDebugTree : Timber.DebugTree() {
        override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
            var tagUsed = tag
            if (tag != null) {
                val threadName = Thread.currentThread().name
                tagUsed = "<$threadName> $tag"
            }
            super.log(priority, tagUsed, message, t)
        }

        override fun createStackElementTag(element: StackTraceElement): String? {
            return super.createStackElementTag(element) + "(Line " + element.lineNumber + ")"  //日志显示行号
        }
    }

    inner class ReleaseTree : ThreadAwareDebugTree() {
        override fun isLoggable(tag: String?, priority: Int): Boolean {
            return priority != Log.VERBOSE && priority != Log.DEBUG && priority != Log.INFO
        }

        override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
            if (!isLoggable(tag, priority)) return
            super.log(priority, tag, message, t)
        }
    }
}
